### ISSUE TITLE
Introduce a `merge` operation for joining two records.

### ISSUE THREAD

**Issue by vjovanov** [scala-records/scala-records#104]
We have the following possibilities:
1. Expose disjoint fields in the joined record and leave common fields in nested records:

```
  r = Rec {
    def r1 = Rec { def y = 2; };
    def r2 = Rec { def y = 3; };
    def x = 1;
    def z = 4;
  }
```

   If somebody tries to access `r.y` we can generate a compile time error message that says look in `r2` `r3`.
1. Just leave two nested records (Slick does something similar):

```
  r = Rec {
    def r1 = Rec { def x = 1; def y = 2; };
    def r2 = Rec { def y = 3; def z = 4; };
  }
```

   Here we could also have an error message if somebody misses a field.
1. Leave both nested records and provide disjoint fields in the joined record:

```
  r = Rec {
    def r1 = Rec { def x = 1; def y = 2; };
    def r2 = Rec { def y = 3; def z = 4; };
    def x = 1;
    def z = 4;
  }
```

**Comment by cvogt** [comment]
Here is what I am imagining for cvogt/records: https://github.com/cvogt/records/issues/2

**Comment by gzm0** [comment]
Crossposting @cvogt's comment from #119:

so one feature I have in compossible is merging of records

```
val r_a: Record[{def name: String; def age: Int}] = Record(name="Chris") ++ Record(age=99)
```

this becomes a problem when field names overlap

```
val r_b: Record[{def age: Long; def age: Int}] = Record(age=101L) ++ Record(age=99)
```

which is a valid Scala type. Preventing these cases when implementing `++` as a macro isn't really viable, if this use case should be supported:

```
def merge[T,S](r1: Record[T], r2: Record[S]): Record[T with S] = r1 ++ r2
```

because here `++` can't know if T and S have name clashes. We can't statically check during lookup either:

```
def lookup[T](r: Record[T with {def age: Int}]) = r.age
lookup(merge(r_b))
```

we can't know if T contains another def age.

So how should we handle that?

One idea would be not only looking up values by field name, but also by type after that. But TypeTags are expensive and do not have referential equality. The latter may not be a problem. Using Fully Qualified names instead would break with path dependent types.

And how should we handle upcasting in this case?

```
val r: Record[{def age: Any}] = Record(age=101L) ++ Record(age=99)
r.age
```

I don't have good answers to these. The best course of action to me right now seems to be deciding on the "best" out of the bad potential behaviors we can think of and well document these.

For example

Alternative 1 (sound, but probably impractical):
- allow overloaded fields and use type tags of the expected return types to find which value to return in case of an overload (not sure using the expected return type actually works well for this)
- live with the performance hit

Alternative 2 (unsound):
- do not use type tags, but something lighter but do the same kind of lookup, e.g. fully qualified names of types and all super types to allow lookup
- live with it being broken for path dependent types

Alternative 3 (sound(?), but runtime exceptions in cases that would pass in 2, but also cases that would be just wrong in 2):
- always fail for overloaded names at compile times when we can and at runtime when we weren't able to find them at compile time, even if Alternative 2 would have worked in that case
- live with runtime exceptions, but at least fail as early as we can

Any better suggestions? Any preferences? I have a slight tendency towards 3. Pushing people to avoid name clashes altogether and fail early and clearly even if at runtime seems better than making them work almost always. If we ever get faster type tags e.g. using scala meta (cc @xeno-by) we may be able to switch to Alternative 1.

**Comment by gzm0** [comment]
I think we definitely should not allow multiple fields with the same name (and different types).

IMO merging the record should have one of the following resolution strategies based on the **static** type:

- Take first
- Fail

I actually propose that we make two different methods to do that (and in the scala-records project, the method would also probably have to be external).

This has a couple of implications:

- The merge method probably has to be a whitebox macro
- "Refined record types" (`Rec[{ def a: Int } with T]`) are not allowed in merge. At a later point, we might introduce judgments that would allow merging a record like that, but that is definitely a second step.

**Comment by cvogt** [comment]
@gzm0
So what "strategy" do you suggest for `println(r.age)`, i.e. member access, "Fail"? What could the api for "Take first" look like?

Why whitebox? Blackbox seems enough to me.

Is disallowing to use `++` on generic records `Record[T]` a reasonable limitation? Seems like a pretty big limitation to me. We would basically disallow generics for merging records.

**Comment by gzm0** [comment]
I suggest the strategy to be part of the merge, not of the access.

The apis could be as follows:

```scala
val x = Rec(a = "foo")
val y = Rec(a = 1)

merge(x, y) // fail, `a` defined twice
mergeFirst(x, y): Rec[ { def a: String } ]
```

> Why whitebox? Blackbox seems enough to me.

Indeed, Blackbox can be enough, depending on the exact semantics. In terms of the names above, `merge` can be implemented as Blackbox (and potentially fail):

```scala
def merge[T, U](x: Rec[T], y: Rec[U]): Rec[T with U]
```

However (as you pointed out in your post), if we pick the first field, we need to strip a member away, so the macro cannot be Blackbox.

> Is disallowing to use ++ on generic records Record[T] a reasonable limitation?

Yes, it is. The goal of records is to have static typing judgments about the available fields and their types on a given member. If such a judgment cannot be made, no record should be created.

**Comment by Ciantic** [comment]
What does merge mean? In TypeScript there is type intersection, and it happily intersects even incompatible types (Auto completion of Atom-typescript list in the example pic):

![image](https://cloud.githubusercontent.com/assets/64731/8691316/bad493f2-2ac9-11e5-960e-ec2c848902c3.png)

One just can't use incompatible type merges e.g. `string & number` them in TypeScript, except setting as undefined. I don't think this issue of one or the either, because there could be another operator that merges incompatible types using Either, depending on the use-case it might be more convenient.

**Comment by gzm0** [comment]
The difference here, is the we are merging values (and types). So a simple example would be:

```scala
val x = Rec(a = 1)
val y = Rec(b = "foo")
val z = merge(a, b)
z: Rec[{ def a: Int, def b: String }] = Rec(a = 1, b = "foo")
```

So a merge must have an actual value as a result (or fail, optimally at compile time). On the other hand, an having an ussatisfiable intersection type is fine since any assignment would fail (in Scala, you *can* write `Int with String` if it makes you happy).

**Comment by Ciantic** [comment]
Yes you are right, though I think I diluted my own point with the TS example. My only point was that there are more than one way to merge, e.g. when it hits incompatible types it wraps the types (and the values) in `Either`.

**Comment by cvogt** [comment]
@gzm0 alright, let's go with disallowing merges on abstract types. Not sure we get away with it forever but we might and we can always loosen the screws later.

Also, how about instead of having a whitebox `mergeFirst`, we have only a blackbox `merge` and a whitebox "remove field operator". Those two subsume `mergeFirst`. E.g. 
```
val r: Record[{def age: Long; ...}] = Record(age=99L, ...) ++ (Record(age=99, ...).remove(_.age))
```
(remove syntax only as an example of course)

**Comment by gzm0** [comment]
Yes, that seems reasonable. The nice thing is, we don't even need the remove field as of now, since it can be done with an upcast (of course having appropriate projections is better, but we don't need them immediately).

**Comment by vjovanov** [comment]
I think failing is not a good idea. Clashes happen quite often and it is very ugly when you have to rename fields you really do not care about. It can lead to long lines of useless code. 

Why wouldn't we do what SQL does: i) when there is a conflict you can access it by explicitly saying do you want left or right version or, ii) simply use `Either` for all conflicts on the field level. The base record has all non-conflicted things. 

For the sake of type errors I would go with `Either`.

**Comment by cvogt** [comment]
@vjovanov That's a good way to look at it. I am warming up to the idea of Either. What bugs me though is that I think a blackbox macro may be more important than "Either merging behavior" for the majority of people. I think we should definitely have a blackbox merge that is as good as possible and then we can have a different one that does the either variant. WDYT?

I am using ++ as the blackbox merge operator in compossible, because shapeless uses that. I have & as well, but warmed up to ++ in preparation to my scala days talk. Not sure what shapeless does in case of conflicts. We may want to follow shapeless' naming choices. Maybe +++ could be the other operator.

BTW, what are you planning to do for merging twice. Either[Either[_],_] or Either[_,Either[_]] or something else?

**Comment by vjovanov** [comment]
Having both a blackbox and a whitebox version of the operator would be good for the library. Not sure about `Either[Either[,],]` it is pretty inconvenient. Maybe a tuple or an HList would fit nicer into the picture ,or even nested records.
